File: LeetCode/905.Sort_Array_By_Parity/905.Sort_Array_By_Parity.swift
class Solution {
    func sortArrayByParity(_ nums: [Int]) -> [Int] {
        // two pointers,
        // i: traverse entire nums
        // j: only shift when num is even

        var nums = nums
        let n = nums.count
        var j = 0

        for (i, num) in nums.enumerated() {
            if num % 2 == 0 {
                nums.swapAt(i, j)
                j += 1
            }
        }

        return nums
    }
}

File: LeetCode/905.Sort_Array_By_Parity/905.Sort_Array_By_Parity.py
#!/usr/bin/env python3


class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        # two pointer
        # one point to even, one point to every element
        n = len(nums)
        # j point to even
        j = 0
        for i in range(n):
            if nums[i] % 2 == 0:
                nums[i], nums[j] = nums[j], nums[i]
                j += 1

        return nums
